### ISSUE TITLE
loompy write speed -- 4 MB per minute?

### ISSUE THREAD

**Issue by jeff-regier** [linnarsson-lab/loompy#48]
I'm attempting to convert a large scRNA-seq dataset from a sparse matrix format to the loom format using loompy 2. While `loompy.create` is running, the output file grows at a rate of roughly 4 MB / minute. Is that the expected speed, or am I doing something inefficiently? The dataset (on disk, in a sparse COO format) is about 4 GB on disk. At this rate, the conversion should take at least 16 hours---longer if the output file is substantially larger that the input file. Excluding imports, my current code is just three lines:

```
f = h5py.File("1M_neurons_filtered_gene_bc_matrices_h5.h5")
X = csc_matrix((f["mm10"]["data"], f["mm10"]["indices"], f["mm10"]["indptr"]))
loompy.create("test.loom", X, {}, {})
```

Are there any flags or options that I could be using to do this conversion more efficiently?

Thanks!

**Comment by slinnarsson** [comment]
That doesn't sound right. I'll look into the issue (maybe CSC is inefficient compared to CSR?) but in the meantime, here's some code that worked for me and finished in half an hour:

```python
f = h5py.File("/Users/sten/10X_datasets/1.3M_neurons/1M_neurons_filtered_gene_bc_matrices_h5.h5")
genes = f["/mm10/gene_names"][:].astype("unicode")
accessions = f["/mm10/genes"][:].astype("unicode")
indptr = f["/mm10/indptr"][:]

def read_n(i, n):
    if i + n > f['/mm10/indices'].shape[0]:
        n = f['/mm10/indices'].shape[0] - i
    cells = np.zeros((genes.shape[0], n))
    for ix in range(i, i + n):
        a = indptr[ix]
        b = indptr[ix + 1]
        cells[f['/mm10/indices'][a:b], ix - i] = f["/mm10/data"][a:b]
    return cells

i = 0
ds = None
while True:
    cells = read_n(i, 1000)
    print(i)
    if ds is None:
        loompy.create("10x_1.3M_cells.loom", cells, {"Gene": genes}, {"CellID": np.fromiter(range(i, i + 1000), dtype='int')})
        ds = loompy.connect("10x_1.3M_cells.loom")
    else:
        ds.add_columns(cells, {"CellID": np.fromiter(range(i, i + 1000), dtype='int')})
    i = i + 1000
```

You'll have to change the file path of course. Also, I didn't bother about the 10X cell barcodes, but if you want them, you'll have to pass them instead of the `np.fromiter(range(i, i + 1000), dtype='int')` both when creating the initial file and when adding columns.

**Comment by slinnarsson** [comment]
Do you have enough RAM? The result of `f["mm10"]["indptr"]` is a very large array, so maybe you hit the limit of physical RAM and the computer starts [thrashing](https://en.wikipedia.org/wiki/Thrashing_(computer_science))?

Note that in my code above I avoided reading the full `indptr` array into RAM, which might explain the difference in performance.

**Comment by slinnarsson** [comment]
Ok indeed there was a severe performance bug. I'm not sure I understand why it happened, but I found a fix. Loompy used to append sparse matrices 64 columns at a time. For a 480 MB test dataset this would take 1 hour and 14 minutes and result in a 3.22 GB (!) output file with exactly the same content as the original. Somehow HDF5 is wasting a lot of space.

Changing the code to write 6400 columns at a time reduced the time to 2 min 5 s and the file size to 338 MB (smaller than the input!).

Evidently there's some weird interaction between write patterns and the efficiency of compression and wasted file space. 

Test code for a dataset X of shape (27998, 65179) in COO format (`float32` elements):

```python
with loompy.connect("/Users/sten/build_20171205/L1_Medulla.loom") as ds:
    X = ds.sparse()  # 3 min 51 s
    loompy.create("test6.loom", X, ds.ra, ds.ca) # 2 min 5 s
```

Writing the matrix as `int16` reduced the time further to 1 min 45 s and the size to 280 MB:

```python
Y = sparse.coo_matrix((X.data.astype('int16'), (X.row, X.col)), shape=X.shape, dtype='int16')
with loompy.connect("/Users/sten/build_20171205/L1_Medulla.loom") as ds:
    loompy.create("test8.loom", Y, ds.ra, ds.ca)
```

I will prepare a release with this fix. Thanks for finding the bug!

**Comment by jeff-regier** [comment]
Hi Sten, Thanks for looking into it, and for the script! I actually like the that script more than loading everything into memory anyway. I was running on a "big memory" machine, so there wasn't any thrashing---CPU was pegged at 100%. But nicer to have code that runs on a machine with a normal amount of RAM too.

**Comment by slinnarsson** [comment]
Fixed in [2.0.9](https://github.com/linnarsson-lab/loompy/releases/tag/2.0.9).
